From a8d8e8ed52e3cae586d02b9d67cc3d3bf35a97fb Mon Sep 17 00:00:00 2001
From: Tanbir Jishan <tantalising007@gmail.com>
Date: Wed, 29 Jun 2022 19:26:11 +0000
Subject: [PATCH] kcms/colors: Show last-used custom accent color when clicking
 custom color checkbox

Right now it opens the color picker dialog with the current first accent color.
Remembering the user's prior custom accent color (if any) is nicer.

BUG: 455894
FIXED-IN: 5.26
---
 kcms/colors/colors.cpp                   | 13 +++++++++++++
 kcms/colors/colors.h                     |  5 +++++
 kcms/colors/colorssettings.kcfg          |  4 ++++
 kcms/colors/package/contents/ui/main.qml |  6 ++++--
 4 files changed, 26 insertions(+), 2 deletions(-)

diff --git a/kcms/colors/colors.cpp b/kcms/colors/colors.cpp
index 32daa35e7..a0d3e34e9 100644
--- a/kcms/colors/colors.cpp
+++ b/kcms/colors/colors.cpp
@@ -147,6 +147,19 @@ void KCMColors::setAccentColorFromWallpaper(bool boolean)
     Q_EMIT settingsChanged();
 }
 
+QColor KCMColors::lastUsedCustomAccentColor() const
+{
+    return colorsSettings()->lastUsedCustomAccentColor();
+}
+void KCMColors::setLastUsedCustomAccentColor(const QColor &accentColor)
+{
+    if (accentColor == QColor(Qt::transparent)) {
+        return;
+    }
+    colorsSettings()->setLastUsedCustomAccentColor(accentColor);
+    Q_EMIT lastUsedCustomAccentColorChanged();
+    Q_EMIT settingsChanged();
+}
 bool KCMColors::downloadingFile() const
 {
     return m_tempCopyJob;
diff --git a/kcms/colors/colors.h b/kcms/colors/colors.h
index 907fb0711..bd28a182e 100644
--- a/kcms/colors/colors.h
+++ b/kcms/colors/colors.h
@@ -42,6 +42,7 @@ class KCMColors : public KQuickAddons::ManagedConfigModule
     Q_PROPERTY(ColorsSettings *colorsSettings READ colorsSettings CONSTANT)
     Q_PROPERTY(bool downloadingFile READ downloadingFile NOTIFY downloadingFileChanged)
     Q_PROPERTY(QColor accentColor READ accentColor WRITE setAccentColor NOTIFY accentColorChanged)
+    Q_PROPERTY(QColor lastUsedCustomAccentColor READ lastUsedCustomAccentColor WRITE setLastUsedCustomAccentColor NOTIFY lastUsedCustomAccentColorChanged)
     Q_PROPERTY(bool accentColorFromWallpaper READ accentColorFromWallpaper WRITE setAccentColorFromWallpaper NOTIFY accentColorFromWallpaperChanged)
 
 public:
@@ -68,6 +69,10 @@ public:
     void resetAccentColor();
     Q_SIGNAL void accentColorChanged();
 
+    QColor lastUsedCustomAccentColor() const;
+    void setLastUsedCustomAccentColor(const QColor &accentColor);
+    Q_SIGNAL void lastUsedCustomAccentColorChanged();
+
     bool accentColorFromWallpaper() const;
     void setAccentColorFromWallpaper(bool boolean);
     Q_SIGNAL void accentColorFromWallpaperChanged();
diff --git a/kcms/colors/colorssettings.kcfg b/kcms/colors/colorssettings.kcfg
index 06bdc1c31..700d803d8 100644
--- a/kcms/colors/colorssettings.kcfg
+++ b/kcms/colors/colorssettings.kcfg
@@ -13,6 +13,10 @@
       <label>Accent color</label>
       <default>transparent</default>
     </entry>
+    <entry name="lastUsedCustomAccentColor" key="LastUsedCustomAccentColor" type="Color">
+      <label>The last used custom accent color before user switched to wallpaper generated color or default color</label>
+      <default>transparent</default>
+    </entry>
     <entry name="accentColorFromWallpaper" key="accentColorFromWallpaper" type="Bool">
       <label>Whether accent color from wallpaper should be applied</label>
       <default>false</default>
diff --git a/kcms/colors/package/contents/ui/main.qml b/kcms/colors/package/contents/ui/main.qml
index 83eac29d9..00d5b133f 100644
--- a/kcms/colors/package/contents/ui/main.qml
+++ b/kcms/colors/package/contents/ui/main.qml
@@ -174,7 +174,7 @@ KCM.GridViewKCM {
                     onToggled: {
                         if (checked) {
                             kcm.accentColorFromWallpaper = false;
-                            kcm.accentColor = colorRepeater.model[0]
+                            kcm.accentColor = Qt.colorEqual(kcm.lastUsedCustomAccentColor, "transparent") ? colorRepeater.model[0] : kcm.lastUsedCustomAccentColor
                         }
                     }
                 }
@@ -237,6 +237,7 @@ KCM.GridViewKCM {
                             onToggled: {
                                 kcm.accentColorFromWallpaper = false;
                                 kcm.accentColor = modelData
+                                kcm.lastUsedCustomAccentColor = modelData
                                 checked = Qt.binding(() => Qt.colorEqual(kcm.accentColor, modelData));
                             }
                         }
@@ -254,9 +255,10 @@ KCM.GridViewKCM {
                         title: i18n("Choose custom accent color")
                         // User must either choose a colour or cancel the operation before doing something else
                         modality: Qt.ApplicationModal
-                        color: kcm.accentColor
+                        color: Qt.colorEqual(kcm.lastUsedCustomAccentColor, "transparent") ? kcm.accentColor : kcm.lastUsedCustomAccentColor
                         onAccepted: {
                             kcm.accentColor = colorDialog.color
+                            kcm.lastUsedCustomAccentColor = colorDialog.color
                             kcm.accentColorFromWallpaper = false
                         }
                     }
-- 
GitLab

